File: 0072/dp.py
"""
  |- h o r s e
--------------
 -|0
 r|  1
 o|    1 2
 s|        2 3
"""


class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        old_cost = [i for i in range(len(word1) + 1)]
        for idx2 in range(len(word2)):
            new_cost = [None for i in range(len(word1) + 1)]
            new_cost[0] = old_cost[0] + 1   # delete
            for idx1 in range(len(word1)):
                cost_u = old_cost[idx1+1]
                cost_l = new_cost[idx1]
                cost_ul = old_cost[idx1]
                # identical
                if word2[idx2] == word1[idx1]:
                    new_cost[idx1+1] = cost_ul
                else:
                    # exchange
                    if cost_ul < cost_l and cost_ul < cost_u:
                        new_cost[idx1+1] = cost_ul + 1
                    # insert
                    elif cost_l < cost_u:
                        new_cost[idx1+1] = cost_l + 1
                    # delete
                    else:
                        new_cost[idx1+1] = cost_u + 1
            old_cost = new_cost
        return old_cost[-1]


if __name__ == '__main__':
    sol = Solution()
    w1 = 'horse'
    w2 = 'ros'
    got = sol.minDistance(w1, w2)
    assert got == 3

    w1 = 'intention'
    w2 = 'execution'
    got = sol.minDistance(w1, w2)
    assert got == 5

File: 0072/dp.cpp
#include <algorithm>
#include <cstdlib>
#include <string>
#include <vector>

#include "gtest/gtest.h"

using namespace std;

/**
 *   - h o r s e
 * - 0 1 2 3 4 5
 * r 1 1 2 2 3 4
 * o 2 2 1 2 3 4
 * s 3 2 2 1 2 3
 */

class Solution {
 public:
  int minDistance(string word1, string word2) {
    const size_t m = word2.size(), n = word1.size();
    vector<int> prev(n+1), curr(n+1);
    for (size_t j = 0; j <= n; ++j) { prev[j] = j; }
    for (size_t i = 1; i <= m; ++i) {
      curr[0] = i;
      const char& c2 = word2[i-1];
      for (size_t j = 1; j <= n; ++j) {
        if (word1[j-1] == c2) {
          curr[j] = prev[j-1];
        } else {
          curr[j] = min({prev[j-1] + 1, prev[j] + 1, curr[j-1] + 1});
        }
      }
      prev.swap(curr);
    }
    return prev[n];
  }
};

int main() {
  Solution sol;
  string word1, word2;

  word1 = "horse", word2 = "ros";
  EXPECT_EQ(3, sol.minDistance(word1, word2));

  word1 = "intention", word2 = "execution";
  EXPECT_EQ(5, sol.minDistance(word1, word2));

  exit(EXIT_SUCCESS);
}
